hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/ProvidesHoodieConfig.scala

```diff
@@ -89,6 +89,62 @@ trait ProvidesHoodieConfig extends Logging {
       defaultOpts = defaultOpts, overridingOpts = overridingOpts)
   }
 
+  /**
+   * Get the insert operation.
+   * See if we are able to set bulk insert, else use deduceOperation
+   */
+  private def getOperation(isPartitionedTable: Boolean,
+                           isOverwritePartition: Boolean,
+                           isOverwriteTable: Boolean,
+                           insertModeSet: Boolean,
+                           dropDuplicate: Option[String],
+                           enableBulkInsert: Option[String],
+                           isInsertInto: Boolean,
+                           isNonStrictMode: Boolean,
+                           hasPrecombineColumn: Boolean): String = {
+    val notSetToNonStrict = !insertModeSet || isNonStrictMode
```

Comment:
if we pass in the incoming params, would that make things easier. 
for eg
```
val insertModeOverridden = params.containsKey("hoodie.sql.insert.mode") 
```

something like this